src/backends/plonky2/primitives/merkletree_circuit.rs

```diff
@@ -0,0 +1,484 @@
+            extra_bits[..n_extra_bits].to_vec()
+            // Note: ideally we would do:
+            //     let extra_bits = builder.split_le(key[n_complete_field_elems], n_extra_bits);
+            // and directly get the extra_bits, but the `split_le` method
+            // returns the wrong bits, so currently we get the entire array of
+            // bits and crop it at the desired n_extra_bits amount.
+        } else {
+            vec![]
+        };
+        [path, extra_bits].concat()
+    }
+
+    fn kv_hash_target(
+        builder: &mut CircuitBuilder<F, D>,
+        key: &Vec<Target>,
+        value: &Vec<Target>,
+    ) -> HashOutTarget {
+        let inputs: Vec<Target> = [key.clone(), value.clone(), vec![builder.one()]].concat();
+        builder.hash_n_to_hash_no_pad::<PoseidonHash>(inputs)
+    }
+}
+
+#[cfg(test)]
+pub mod tests {
+    use plonky2::plonk::{circuit_builder::CircuitBuilder, circuit_data::CircuitConfig};
+    use std::collections::HashMap;
+
+    use super::*;
+    use crate::backends::plonky2::basetypes::hash_value;
+    use crate::backends::plonky2::basetypes::C;
+    use crate::backends::plonky2::primitives::merkletree::*;
+
+    #[test]
+    fn test_keypath() -> Result<()> {
+        test_keypath_opt::<10>()?;
+        test_keypath_opt::<16>()?;
+        test_keypath_opt::<32>()?;
+        test_keypath_opt::<40>()?;
+        test_keypath_opt::<64>()?;
+        test_keypath_opt::<128>()?;
+        test_keypath_opt::<130>()?;
+        test_keypath_opt::<250>()?;
+        test_keypath_opt::<256>()?;
+        Ok(())
+    }
+
+    fn test_keypath_opt<const MD: usize>() -> Result<()> {
+        for i in 0..5 {
+            let config = CircuitConfig::standard_recursion_config();
+            let mut builder = CircuitBuilder::<F, D>::new(config);
+            let mut pw = PartialWitness::<F>::new();
+
+            let key = Value::from(hash_value(&Value::from(i)));
+            let expected_path = keypath(MD, key)?;
+
+            // small circuit logic to check
+            // expected_path_targ==keypath_target(key_targ)
+            let expected_path_targ: Vec<BoolTarget> = (0..MD)
+                .map(|_| builder.add_virtual_bool_target_safe())
+                .collect();
+            let key_targ = builder.add_virtual_targets(VALUE_SIZE);
+            let computed_path_targ =
+                MerkleProofCircuit::<MD>::keypath_target(&mut builder, &key_targ);
+            for i in 0..MD {
+                builder.connect(computed_path_targ[i].target, expected_path_targ[i].target);
+            }
+
+            // assign the input values to the targets
+            pw.set_target_arr(&key_targ, &key.0.to_vec())?;
+            for i in 0..MD {
+                pw.set_bool_target(expected_path_targ[i], expected_path[i])?;
+            }
+
+            // generate & verify proof
+            let data = builder.build::<C>();
+            let proof = data.prove(pw)?;
```

Comment:
Is there any functionality in plonky2 to verify the constraints without computing the proof?
Some sort of circuit debugger that just goes through all constraints and verifies that they pass after the witness assignment?

I'm asking this because we could use that in tests to run them much faster.

Reply:
Oh I see!  That's unfortunate then.